comp/core/autodiscovery/scheduler/controller.go

```diff
@@ -0,0 +1,204 @@
+type Controller struct {
+	// m protects all fields in this struct.
+	m sync.Mutex
+
+	// activeSchedulers is the set of schedulers currently subscribed to configs.
+	activeSchedulers map[string]Scheduler
+
+	// scheduledConfigs contains the set of configs that have been scheduled
+	// via the schedulerController, but not subsequently unscheduled.
+	scheduledConfigs map[string]*integration.Config
+
+	// ConfigStateStore contains the desired state of configs
+	ConfigStateStore *ConfigStateStore
+
+	// a workqueue to process the config events
+	queue workqueue.RateLimitingInterface
+
+	started     bool
+	stopChannel chan struct{}
+}
+
+type workItem struct {
+	digest Digest
+}
+
+// NewController inits a scheduler controller
+func NewController() *Controller {
+	schedulerController := Controller{
+		scheduledConfigs: make(map[string]*integration.Config),
+		activeSchedulers: make(map[string]Scheduler),
+		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "Controller"),
+		stopChannel:      make(chan struct{}),
+		ConfigStateStore: NewConfigStateStore(),
+	}
+	schedulerController.start()
+	return &schedulerController
+}
+
+func (ms *Controller) start() {
+	ms.m.Lock()
+	if ms.started {
+		return
+	}
+	ms.started = true
+	ms.m.Unlock()
+	go wait.Until(ms.worker, time.Second, ms.stopChannel)
+}
+
+// Register a new scheduler to receive configurations.
+// Previously scheduled configurations that have not subsequently been
+// unscheduled can be replayed with the replayConfigs flag.  This replay occurs
+// immediately, before the AddScheduler call returns.
+func (ms *Controller) Register(name string, s Scheduler, replayConfigs bool) {
+	ms.m.Lock()
+	if _, ok := ms.activeSchedulers[name]; ok {
+		log.Warnf("Scheduler %s already registered, overriding it", name)
+	}
+	ms.activeSchedulers[name] = s
+	ms.m.Unlock()
+
+	// if replaying configs, replay the currently-scheduled configs; note that
+	// this occurs under the protection of `ms.m`, so no config may be double-
+	// scheduled or missed in this process.
+	if replayConfigs {
+		configStates := ms.ConfigStateStore.List()
+
+		configs := make([]integration.Config, 0, len(configStates))
+		for _, config := range configStates {
+			if config.desiredState == Scheduled {
+				configs = append(configs, *config.config)
+			}
+		}
+		s.Schedule(configs)
+	}
+}
+
+// Deregister a scheduler in the schedulerController to dispatch to
+func (ms *Controller) Deregister(name string) {
+	ms.m.Lock()
+	defer ms.m.Unlock()
+	if _, ok := ms.activeSchedulers[name]; !ok {
+		log.Warnf("Scheduler %s no registered, skipping", name)
+		return
+	}
+	delete(ms.activeSchedulers, name)
+}
+
+// ApplyChanges add configDigests to the workqueue
+func (ms *Controller) ApplyChanges(digests []Digest) {
+	for _, configDigest := range digests {
+		ms.queue.Add(workItem{digest: configDigest})
+	}
+}
+
+// Schedule for test only, actual scheduling should be done by the autoconfig.applyChanges
```

Comment:
Two questions:
- The tests are in the same package, why do we need to be public?
- Are tests testing un-used APIs useful?